design/boto-design.md

```diff
@@ -0,0 +1,101 @@
+# Co-Existing with Boto design
+
+## Problem Description
+
+Existing code bases where paths are scattered throughout the codebase (which is rather common for DE/DS code) have a hard time incorporating the lakeFS S3 gateway for several reasons:
+1. Clients (namely, boto3 and pandas) don't support multiple S3 endpoints with differing credentials. This makes it hard to change only a subset of the data to be in lakeFS because the client, once initialized, can talk to either lakeFS or S3 but not both.
+2. Once data is migrated from S3 to lakeFS, even if the repo has the same name as the previous bucket, it's still not a drop-in replacement, because in lakeFS you'd also need to specify the version ref (branch/commit/tag) you're using, as a prefix.
+
+Solving 1 is hard because it requires a big infrastructure change: instead of using a single initialized client (sometimes it's a default or global one), now multiple clients need to be introduced (one for S3, one for lakeFS) with custom logic for when to use which one.
+Solving 2 is hard because it involves changing many paths scattered throughout the code base: some are dynamically generated by the code, some come from env vars or configuration files - it's not always easy to ensure coverage.
+
+## Goals
+* Make it easy for new users to transfer existing data to lakefs
+* Isolation - Users should be able to separate access to bucket/prefix between lakefs and s3
+* Users should do the minimal set of changes possible in code to set lakefs
+
+## Non Goals
+* Provide new features that weren't previously possible through lakefs
+
+## Value Propositions
+When a user with an existing python code that uses boto3 client to access s3 wants to move only part of its buckets/prefixes to lakefs, botwo allows using lakefs and s3 together with minimum user's code changes.
+
+## Proposed Design
```

Comment:
Please add a picture here, it should make things easier to understand.

Reply:
Didn't understand from the design how do we manage to handle all boto requests? do hard code each method boto currently support? do we dynamically generate them? it is important to understand this part to know how a call to our client on a method that boto holds, implemented after we published the package behaves.

Reply:
It will be published to pypi as a package. 
The implementation will be such that we will always be sync with boto latest version because we dynamically generate the s3 api functions in run time according to boto.

Reply:
So we will not require any specific version of boto as dependency of the 'botwo' package?